--- linux-3.7.0-gentoo/drivers/platform/x86/fujitsu-tablet.c	2012-12-11 05:30:57.000000000 +0200
+++ linux-3.7.0-gentoo.orig/drivers/platform/x86/fujitsu-tablet.c	2012-12-12 22:54:38.948061567 +0200
@@ -72,22 +72,22 @@
 };
 
 static unsigned short keymap_Lifebook_U810[KEYMAP_LEN] __initdata = {
-	KEY_RESERVED,
-	KEY_RESERVED,
+	KEY_F15,
+	KEY_F16,
 	KEY_RESERVED,
 	KEY_RESERVED,
 	KEY_PROG1,
 	KEY_PROG2,
-	KEY_DIRECTION,
+	KEY_PROG3,
 	KEY_RESERVED,
+	KEY_BRIGHTNESSUP,
+	KEY_BRIGHTNESSDOWN,
+	KEY_PAGEUP,
+	KEY_PAGEDOWN,
 	KEY_RESERVED,
 	KEY_RESERVED,
-	KEY_UP,
-	KEY_DOWN,
-	KEY_RESERVED,
-	KEY_RESERVED,
-	KEY_LEFTCTRL,
-	KEY_LEFTALT
+	KEY_RIGHTALT,
+	KEY_RIGHTCTRL
 };
 
 static unsigned short keymap_Stylistic_Tseries[KEYMAP_LEN] __initdata = {
@@ -158,6 +158,7 @@
 
 static void fujitsu_send_state(void)
 {
+	return;
 	int state;
 	int dock, tablet_mode;
 
@@ -190,6 +191,8 @@
 		msleep(20);
 
 	fujitsu_send_state();
+
+	fujitsu.prev_keymask = 2;
 }
 
 static int __devinit input_fujitsu_setup(struct device *parent,
@@ -256,6 +259,8 @@
 	keymask  = fujitsu_read_register(0xde);
 	keymask |= fujitsu_read_register(0xdf) << 8;
 	keymask ^= 0xffff;
+	keymask |= (fujitsu_read_register(0xdd) & 0x01) + 1;
+//	printk("TABLET: 0xdd:%d 0xde:%d 0xdf:%d keymask:%d prev_keymask:%d\n",fujitsu_read_register(0xdd),fujitsu_read_register(0xde),fujitsu_read_register(0xdf),keymask,fujitsu.prev_keymask);
 
 	changed = keymask ^ fujitsu.prev_keymask;
 	if (changed) {
@@ -264,11 +269,16 @@
 		for_each_set_bit(i, &changed, KEYMAP_LEN) {
 			keycode = fujitsu.config.keymap[i];
 			pressed = keymask & changed & BIT(i);
+//			printk("TABLET: keycode:%d changed: %d pressed:%d\n",keycode,changed,pressed);
 
 			if (pressed)
 				input_event(fujitsu.idev, EV_MSC, MSC_SCAN, i);
 
 			input_report_key(fujitsu.idev, keycode, pressed);
+			if (pressed != changed) {
+				input_sync(fujitsu.idev);
+				input_report_key(fujitsu.idev, keycode, 0);
+			}
 			input_sync(fujitsu.idev);
 		}
 	}
